blog/2024-12-06-deadlock-gadget-lfx-mentorship.md

```diff
@@ -0,0 +1,144 @@
+We have come full circle ending up with the threads being stuck in an infinite wait cycle due to mutex lock order inversions, thus hanging the process.
+
+:::tip
+
+The diagram above represents a directed graph. Keep reading to learn how we use it to detect these deadlocks.
+
+:::
+
+You can simulate a deadlock in code using the following C++ program adapted from [BCC tools](https://github.com/iovisor/bcc/blob/master/tools/deadlock_example.txt#L185):
+
+```cpp
+#include <chrono>
+#include <iostream>
+#include <mutex>
+#include <thread>
+
+std::mutex global_mutex1;
+std::mutex global_mutex2;
+
+int main(void) {
+    static std::mutex static_mutex3;
+    std::mutex local_mutex4;
+    
+    std::cout << "sleeping for a bit to allow trace to attach..." << std::endl;
+    std::this_thread::sleep_for(std::chrono::seconds(10));
+    std::cout << "starting program..." << std::endl;
+    
+    auto t1 = std::thread([] {
+    std::lock_guard<std::mutex> g1(global_mutex1);
+    std::lock_guard<std::mutex> g2(global_mutex2);
+    });
+    
+    auto t2 = std::thread([] {
+    std::lock_guard<std::mutex> g2(global_mutex2);
+    std::lock_guard<std::mutex> g3(static_mutex3);
+    });
+    
+    auto t3 = std::thread([&local_mutex4] {
+    std::lock_guard<std::mutex> g3(static_mutex3);
+    std::lock_guard<std::mutex> g4(local_mutex4);
+    });
+    
+    auto t4 = std::thread([&local_mutex4] {
+    std::lock_guard<std::mutex> g4(local_mutex4);
+    std::lock_guard<std::mutex> g1(global_mutex1);
+    });
+    
+    // Wait for the threads to finish execution..
+    // WARNING! You might never reach this part of the code!
+    t1.join();
+    t2.join();
+    t3.join();
+    t4.join();
+    
+    std::cout << "sleeping to allow trace to collect data..." << std::endl;
+    std::this_thread::sleep_for(std::chrono::seconds(5));
+    std::cout << "done!" << std::endl;
+}
+```
+
+## The Gadget
+
+Inspektor Gadget recently unlocked support for uprobe based gadgets, as part of another LFX Mentorship project by fellow inspektor [Tianyi Liu](https://github.com/i-Pear) in the spring term. Read more about it [here](https://www.inspektor-gadget.io/blog/2024/06/supporting-uprobe-based-gadgets-lfx-mentorship-report).
+
+Using uprobes, we attach to the `pthread_mutex_lock` and `pthread_mutex_unlock` methods from the `pthread` library ie. POSIX threads (now integrated into the GNU’s standard `libc` library), as follows:
+
+```c
+SEC("uprobe/libc:pthread_mutex_lock")
+int BPF_UPROBE(trace_uprobe_mutex_lock, void *mutex_addr)
+{
+    // Intercept it as you want!
+}
+```
+
+[POSIX threads](https://man7.org/linux/man-pages/man7/pthreads.7.html) is the standard threading library on Unix-like systems, providing low-level control over thread creation, synchronization, and management.
```

Comment:
```suggestion
[POSIX threads](https://man7.org/linux/man-pages/man7/pthreads.7.html) is the standard threading library on Unix-like systems, providing low-level control over thread creation, synchronization and management.
```

Reply:
I have no idea :sweat_smile:

Reply:
Let's keep it then😅? Feels more readable to me..